modal/image.py

```diff
@@ -1299,6 +1299,155 @@ def build_dockerfile(version: ImageBuilderVersion) -> DockerfileSpec:
+                and "uv" in pyproject_toml_content["tool"]
+                and "workspace" in pyproject_toml_content["tool"]["uv"]
+            ):
+                raise InvalidError("uv workspaces are not support")
+
+            if version > "2024.10":
+                # For builder version > 2024.10, modal is mounted at runtime and is not
+                # a requirement in `uv.lock`
+                return
+
+            dependencies = pyproject_toml_content["project"]["dependencies"]
+
+            for group in groups:
+                if (
+                    "dependency-groups" in pyproject_toml_content
+                    and group in pyproject_toml_content["dependency-groups"]
+                ):
+                    dependencies += pyproject_toml_content["dependency-groups"][group]
+
+            for extra in extras:
+                if (
+                    "project" in pyproject_toml_content
+                    and "optional-dependencies" in pyproject_toml_content["project"]
+                    and extra in pyproject_toml_content["project"]["optional-dependencies"]
+                ):
+                    dependencies += pyproject_toml_content["project"]["optional-dependencies"][extra]
+
+            PACKAGE_REGEX = re.compile(r"^[\w-]+")
+
+            def _extract_package(package) -> str:
+                m = PACKAGE_REGEX.match(package)
+                return m.group(0) if m else ""
+
+            if not any(_extract_package(dependency) == "modal" for dependency in dependencies):
+                raise InvalidError(
+                    "Image builder version <= 2024.10 requires modal to be specified in your pyproject.toml file"
+                )
+
+        def build_dockerfile(version: ImageBuilderVersion) -> DockerfileSpec:
+            uv_project_dir_ = os.path.expanduser(uv_project_dir)
+            pyproject_toml = os.path.join(uv_project_dir_, "pyproject.toml")
+
+            uv_root = "/.uv"
+            uv_sync_args = [
+                f"--project={uv_root}",
+                "--no-install-workspace",  # Do not install the root project or any "uv workspace"
+                "--no-managed-python",  # Use the system python interpreter to create venv
+                "--compile-bytecode",
+            ]
+
+            for group in groups:
+                uv_sync_args.append(f"--group={group}")
+            for extra in extras:
+                uv_sync_args.append(f"--extra={extra}")
+            if extra_options:
+                uv_sync_args.append(extra_options)
+
+            commands = ["FROM base"]
+
+            if uv_version is None:
+                commands.append("COPY --from=ghcr.io/astral-sh/uv:latest /uv /.uv/uv")
+            else:
+                commands.append(f"COPY --from=ghcr.io/astral-sh/uv:{uv_version} /uv /.uv/uv")
+
+            context_files = {}
+
+            _check_pyproject_toml(pyproject_toml, version)
+
+            context_files["/.pyproject.toml"] = pyproject_toml
```

Comment:
jw but why make the context paths "private"?

Reply:
I'm trying to stay consistent with the pattern here:

https://github.com/modal-labs/modal-client/blob/c430972947bf7e471bfece99f3d56657ab1e953b/modal/image.py#L1251